### ISSUE TITLE
About the implementation of gin

### ISSUE THREAD

**Issue by dongZheX** [rusty1s/pytorch_geometric#2863]
From now on, we recommend using our discussion forum (https://github.com/rusty1s/pytorch_geometric/discussions) for general questions.

## ❓ Questions & Help

Hello, thanks for the code.
I have a question about the implementation of gin.

In the benmark/kernel, the sequence of GINCOV is :
 Sequential(
                        Linear(hidden, hidden),
                        ReLU(),
                        Linear(hidden, hidden),
                        ReLU(),
                        BN(hidden)
                    )

but in the example, the sequence of GINCOV is :
Sequential(
                        Linear(dim, dim), 
                        BatchNorm1d(dim), 
                        ReLU(),
                        Linear(dim, dim), 
                        ReLU()
                 )

But in [https://github.com/weihua916/powerful-gnns],
the offical implementation of GIN equals to the following sequence:
Sequential(
                        Linear(dim, dim), 
                        BatchNorm1d(dim), 
                        ReLU(),
                        BatchNorm1d(dim), 
                        Linear(dim, dim), 
                        ReLU()
                 )

could you please tell me the difference of them? or, this is only a mistake.

Looking forward to your reply.

**Comment by rusty1s** [comment]
I don't think there is a real consensus on where to put `BatchNorm` in an MLP. In the end, I think that all implementations perform equally well. Did you do some studies about which implementation works best?

**Comment by dongZheX** [comment]
Thanks.
I have been using the first way of implementation all the time. Today, I firstly notice that there is a difference with the offical implementation.  I thought you do this with the special intention for better performance.

I'll do evaluation for different implementation.  If there is difference between them in performance, I would be happy to share.

**Comment by dongZheX** [comment]
> Hey @dongZheX did you observe any difference in performance ?
It has been a long time, I remember that it had a significant impact on the results of ogbg-pcba.

**Comment by SauravMaheshkar** [comment]
> > Hey @dongZheX did you observe any difference in performance ?
> 
> It has been a long time, I remember that it had a significant impact on the results of ogbg-pcba.

Huh, interesting I'll run some experiments on my end too. Oh btw I think we can close this issue now @dongZheX ?

**Comment by SauravMaheshkar** [comment]
Documenting some results here : [Does the position of BatchNorm matter in Graph Isomorphism Networks (GIN)](https://wandb.ai/graph-neural-networks/GIN/reports/Does-the-position-of-BatchNorm-matter-in-Graph-Isomorphism-Networks-GIN---Vmlldzo1MDkwMTM3)

**Comment by rusty1s** [comment]
That's pretty cool :) Do you think we need to change anything on our end based on your results?

**Comment by SauravMaheshkar** [comment]
Well so far it seems like having a BatchNorm after each activation function helps however I don't want to make hasty statements based on smaller datasets. I have been working on scaling these experiments to use the datasets from the [Long Range Graph Benchmarks](https://github.com/vijaydwivedi75/lrgb) instead. However it seems like GIN isn't intricate enough to deal with datasets of that size, or maybe my implementation of the test accuracy is faulty. 

Some help with the code would be appreciated. I'm using the following function to calculate the test accuracy

```python
def multilabel_accuracy(
    model: torch.nn.Module,
    data: Data,
    test_dataloader: Iterable,
    threshold: float = 0.5,
) -> float:
    """
    Calculate the accuracy on the test set for multilabel classification

    Args:
        model(torch.nn.Module): Model
        data(Data): Data object
        test_dataloader(Iterable): Test Dataloader
        threshold(float): Threshold for prediction, defaults to 0.5

    Returns:
        float: Test Accuracy
    """
    correct: int = 0
    total_instances: int = 0
    with torch.no_grad():
        predictions = model(data)
        predicted_labels = predictions > threshold
        comparison = predicted_labels == data.y
        correct += comparison.all(dim=1).sum().item()
        total_instances += data.y.size(0)
    return correct / total_instances
```

The dataset being used is the LRGBDataset from PyG. If perhaps you could also review the [training script](https://github.com/SauravMaheshkar/Graph-Property-Prediction/blob/main/src/engine.py)

**Comment by rusty1s** [comment]
For evaluation on multi-label, I think you may also take a look at `torchmetrics` (e.g., AUROC, AUPRC, etc). Usually, it is not a good idea to pre-define the threshold without any calibration.

**Comment by rusty1s** [comment]
Does the same hold true if you are using https://torchmetrics.readthedocs.io/en/stable/classification/auroc.html#torchmetrics.classification.MultilabelAUROC?

**Comment by SauravMaheshkar** [comment]
Hey @rusty1s, using MultiLabel AUROC finally led to a stable and coherent training pipeline 🤗 . I just trained using all the various possible BatchNorm positions and have the results.


![Screenshot 2023-08-29 at 1 32 32 PM](https://github.com/pyg-team/pytorch_geometric/assets/61241031/9c5de4c1-4d07-4db4-8ad6-eca5c13925e3)


Although the training looks very similar for all three variants there is obvious improvements while using "Sequential" BatchNorms i.e. a BatchNorm after ReLU in the `GINConv` layer. The same has been observed using smaller datasets from TUDatasets as well. 

The full comparison and benchmark can be found at this [Weights & Biases Report](https://wandb.ai/graph-neural-networks/GIN/reports/Does-the-position-of-BatchNorm-matter-in-Graph-Isomorphism-Networks-GIN---Vmlldzo1MDkwMTM3).

Changing the current GIN implementation to reflect sequential batchnorms is relatively easy. The way I do it in my codebase is

```python
GINConv(
    nn.Sequential(
        nn.Linear(num_features, hidden),
        nn.ReLU(),
        nn.BatchNorm1d(hidden)
        if batchnorm in ["first", "sequential"]
        else nn.Identity(),
        nn.Linear(hidden, hidden),
        nn.ReLU(),
        nn.BatchNorm1d(hidden)
        if batchnorm in ["last", "sequential"]
        else nn.Identity(),
    ),
    train_eps=use_eps,
)
```

I'm also more than happy to run more experiments on other datasets if you have some in mind.

**Comment by rusty1s** [comment]
Thanks. Do you mind making the necessary changes for this in `GIN`?

**Comment by SauravMaheshkar** [comment]
Yes of-course. will open a PR
